ledger/heavy/executor/jetkeeper.go

```diff
@@ -20,143 +20,358 @@ import (
 	"context"
 	"sync"
 
+	"github.com/insolar/insolar/insolar/pulse"
+	"github.com/insolar/insolar/instrumentation/inslogger"
+	"github.com/insolar/insolar/network/storage"
 	"github.com/pkg/errors"
 
 	"github.com/insolar/insolar/insolar"
 	"github.com/insolar/insolar/insolar/jet"
-	"github.com/insolar/insolar/instrumentation/inslogger"
 	"github.com/insolar/insolar/internal/ledger/store"
 )
 
 //go:generate minimock -i github.com/insolar/insolar/ledger/heavy/executor.JetKeeper -o ./ -s _gen_mock.go
 
 // JetKeeper provides a method for adding jet to storage, checking pulse completion and getting access to highest synced pulse.
 type JetKeeper interface {
-	// Add performs adding jet to storage and checks pulse completion.
-	Add(context.Context, insolar.PulseNumber, insolar.JetID) error
+	// AddDropConfirmation performs adding jet to storage and checks pulse completion.
+	AddDropConfirmation(context.Context, insolar.PulseNumber, insolar.JetID, bool) error
+	// AddHotConfirmation performs adding hot confirmation to storage and checks pulse completion.
+	AddHotConfirmation(context.Context, insolar.PulseNumber, insolar.JetID, bool) error
 	// TopSyncPulse provides access to highest synced (replicated) pulse.
 	TopSyncPulse() insolar.PulseNumber
 }
 
-func NewJetKeeper(jets jet.Storage, db store.DB) JetKeeper {
-	return &dbJetKeeper{jetTrees: jets, db: db}
+func NewJetKeeper(jets jet.Storage, db store.DB, pulses pulse.Calculator) JetKeeper {
+	return &dbJetKeeper{
+		jetTrees: jets,
+		db:       db,
+		pulses:   pulses,
+	}
 }
 
 type dbJetKeeper struct {
 	jetTrees jet.Storage
 
+	pulses storage.PulseCalculator
+
 	sync.RWMutex
 	db store.DB
 }
 
-func (jk *dbJetKeeper) Add(ctx context.Context, pulse insolar.PulseNumber, id insolar.JetID) error {
+type jetInfo struct {
+	JetID         insolar.JetID
+	HotConfirmed  []insolar.JetID
+	DropConfirmed bool
+	Split         bool
+}
+
+func (j *jetInfo) addDrop(newJetID insolar.JetID, split bool) error {
+	if j.DropConfirmed {
+		return errors.New("addDrop. try to rewrite drop confirmation. existing: " + j.JetID.DebugString() +
+			", new: " + newJetID.DebugString())
+	}
+	j.Split = split
+	j.DropConfirmed = true
+	j.JetID = newJetID
+
+	return nil
+}
+
+func (j *jetInfo) addHot(newJetID insolar.JetID, split bool) error {
+	if len(j.HotConfirmed) >= 2 {
+		return errors.New("num hot confirmations exceeds 2. existing: " + insolar.JetIDCollection(j.HotConfirmed).DebugString() +
+			", new: " + newJetID.DebugString())
+	}
+	if len(j.HotConfirmed) == 0 {
```

Comment:
You can read and append to a nil slice because append is not mutating in go.
https://goplay.space/#g5zCUknB9zM